### ISSUE TITLE
Datagrid selection: async clrDgSelected with immutable array not working

### ISSUE THREAD

**Issue by Ergnos** [vmware/clarity#4528]
## Describe the bug

When an immutable object is used (in example a ngrx/store state element) the selection breaks when 'item' is pushed to this.current:
```
app.error-handler.ts:19 TypeError: Cannot add property 0, object is not extensible
    at Array.push (<anonymous>)
    at Selection.selectItem (clr-angular.js:11544)
    at Selection.setSelected (clr-angular.js:11581)
    at ClrDatagridRow.set selected [as selected] (clr-angular.js:11815)
    at ClrDatagridRow.toggle (clr-angular.js:11827)
    at ClrDatagridRow_ng_template_5_div_4_Template_input_ngModelChange_1_listener (clr-angular.js:538)
    at executeListenerWithErrorHandling (core.js:21772)
    at wrapListenerIn_markDirtyAndPreventDefault (core.js:21814)
    at SafeSubscriber.schedulerFn [as _next] (core.js:37027)
    at SafeSubscriber.__tryOrUnsub (Subscriber.js:183)
```

## How to reproduce

1.  See stackblitz https://stackblitz.com/edit/clarity-light-theme-v2-cpfinp
2.  Select an item
3.  See console for error message

## Expected behavior
Selection should work with async selection from immutable array.

Solution could be to replace current code:
```
private selectItem(item: T): void {
    this.current.push(item);
    if (this._items.trackBy) {
      // Push selected ref onto array
      const lookup = this._items.all.findIndex(maybe => maybe === item);
      this.prevSelectionRefs.push(this._items.trackBy(lookup, item));
    }
```
to: 
```
private selectItem(item: T): void {
    this.current.concat(item);
    if (this._items.trackBy) {
      // Push selected ref onto array
      const lookup = this._items.all.findIndex(maybe => maybe === item);
      this.prevSelectionRefs.push(this._items.trackBy(lookup, item));
    }
```

Additionally set current(value) can be changed to since otherwise the change event is emitted twice.

```
    set current(value) {
        this.updateCurrent(value, false);
    }
```

## Versions

**App**

* Angular: [8.2.14 & 9.1.1]
* Node: [13.8.0]
* Clarity: [2.4.1 & 3.1.1 (tested with modified clr-angular.js due to issue #4504) ]
* Ngrx/store: [9.1.0]

**Device:**

* Type: [Laptop]
* OS: [Windows 10]
* Browser [Chrome]
* Version [ 81.0.4044.129]

## Additional notes

None

**Comment by skipgeldens** [comment]
When you use an immutable array as selected items you also have to change `deselectItem(indexOfItem)` from:

```
deselectItem(indexOfItem) {
        this.current.splice(indexOfItem, 1);
        if (this._items.trackBy && indexOfItem < this.prevSelectionRefs.length) {
            // Keep selected refs array in sync
            const removedItems = this.prevSelectionRefs.splice(indexOfItem, 1);
            // locked reference is no longer needed (if any)
            this.lockedRefs = this.lockedRefs.filter(locked => locked !== removedItems[0]);
        }
    }
```

To:

```
deselectItem(indexOfItem) {
        this.current = this.current.slice(0, indexOfItem).concat(this.current.slice(indexOfItem+1));
        if (this._items.trackBy && indexOfItem < this.prevSelectionRefs.length) {
            // Keep selected refs array in sync
            const removedItems = this.prevSelectionRefs.splice(indexOfItem, 1);
            // locked reference is no longer needed (if any)
            this.lockedRefs = this.lockedRefs.filter(locked => locked !== removedItems[0]);
        }
    }
```

This is to make sure that the array doesn't get mutated along the way.

**Comment by gnomeontherun** [comment]
Thanks for this, we should be doing this and looks like a straightforward fix. If you'd like to submit a PR for it, I'd be happy to review it, otherwise we'll put it into our backlog.

**Comment by Jinnie** [comment]
We have the fix now. Thanks @skipgeldens 
Note: some backporting will be needed before we can close this issue.

**Comment by Jinnie** [comment]
Merged to v2, v3, v4. Leaving open till a port the double event fixed in v3.
